octodns_azure/__init__.py

```diff
@@ -950,10 +961,57 @@ def _data_for_dynamic(self, azrecord):
 
         return data
 
+    def _required_root_ns_values(self, zone_name, force=False):
+        required_values = self.__required_root_ns_values.get(zone_name, None)
+        if required_values is None or force:
+            required_values = set()
+            if self._check_zone(zone_name):
+                records = self._dns_client.record_sets.list_by_dns_zone
+                for azrecord in records(self._resource_group, zone_name):
```

Comment:
This is an extra call for every zone. Large zones will result in several calls as results are paginated. This is a huge issue with Azure, it often results in 429 rate limiting. So I would strongly prefer avoiding this.

It would be better to populate/cache the NS records during `populate` but I realize that it's called after `_process_desired_zone`. Any chance we can switch that ordering in octodns manager? Or perhaps use processors? Although they don't seem to have access to the desired zone.

Reply:
Sorry I meant to mention that this isn't optimized at all, it's just getting things functional as simply as possible (since i was beating my head against so many false starts/paths.) 

I think the "right" fix here will be to add a helper method that lists the records in a zone and caches it, similar to how `_azure_zones` works now. That pattern is all over octoDNS and should allow `_process_desired_zone` or `populate`  to grab things once whenever the first one runs and the other can just piggy back on the cached version.

Reply:
The rate limiting is so painful that I'd rather this provider not support root NS at all, unless of course we can cache like you said. I'm slightly worried if Azure SDK's record objects are memory-friendly for caching large zones.

But OTOH I think providers can benefit in general by having access to the existing zone during `_process_desired_zone`, which would also make this root NS handling much easier.

Reply:
> I'm slightly worried if Azure SDK's record objects are memory-friendly for caching large zones.

Should be simple enough to convert everything and store the octoDNS records rather than the Azure ones if needed. 

> But OTOH I think providers can benefit in general by having access to the existing zone during `_process_desired_zone`, which would also make this root NS handling much easier. If there's a good reason not to do that, I/we can certainly look into caching.

I'd have to look around, but I guess it's possible to call `_process_desired_zone` after the populate step provided nothing is relying on the order currently.

Reply:
Doesn't look like anything is relying on the order of the 2 operations.

Reply:
Yeah. Thinking about this a bit more, if we move it and then rely on it we'll need to make it explicit when the `_process_*_zone` functions happen and add tests for it since it'll then be a promise that it works that way in the future.

Reply:
I'm not sure of the purpose of `_process_existing_zone` _for_ providers. Any extra processing on top of existing zone can simply be done within the populate phase itself. The only use-case is when we want to process `existing` in `BaseProvider`. If we don't officially open that function up to providers, we can limit the ordering promise between only `populate` and `_process_desired_zone`, which is much simpler than maintaining order of 3 methods.

Reply:
The main thinking was that `populate` is supposed to faithfully reproduce things exactly as they are in the provider currently, not making any "supports" style modifications, beyond what record types are supported. I think that'd mainly come into play when doing a dump from a provider, which just does a straight plan & apply. Even more so when `-lenient` is enabled. 

E.g. we have always have included root NS records in dump thus populate includes them and we wouldn't want to change that. 

The extra method is there to allow providers to make changes after populate saying I know this is what's there, but that's not what I'm going to work with b/c... root NS is the only example of that atm, but it needed solving after populate and it even if this is the only thing that ever needs it it doesn't seem like the end of the world. I guess we could not make it "protected" so that providers don't override and use it.

Reply:
Ok makes sense for having `_process_existing_zone`.

> even if this is the only thing that ever needs it it doesn't seem like the end of the world

It sounded like testing for a promised order of the 3 operations (populate and 2 process methods) may become complex. If we can nail it, I'm all for keeping `_process_existing_zone`. And hopefully we can re-order such that we list zone records just once and preferably without the caching.